File: algorithms/cpp/insertDeleteGetRandomO1/insertDeleteGetRandomO1.cpp
// Source: https://leetcode.com/problems/insert-delete-getrandom-o1/
// Author: Miao Zhang
// Date:   2021-02-05

class RandomizedSet {
public:
    /** Initialize your data structure here. */
    RandomizedSet() {
        
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if (dic_.count(val)) return false;
        dic_[val] = vals_.size();
        vals_.push_back(val);
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        if (!dic_.count(val)) return false;
        int index = dic_[val];
        dic_[vals_.back()] = index;
        dic_.erase(val);
        swap(vals_[index], vals_.back());
        vals_.pop_back();
        return true;
    }
    
    /** Get a random element from the set. */
    int getRandom() {
        int index = rand() % vals_.size();
        return vals_[index];
    }

private:
    unordered_map<int, int> dic_;
    vector<int> vals_;
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: algorithms/python/insertDeleteGetRandomO1/insertDeleteGetRandomO1.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/insert-delete-getrandom-o1/
# Author: Miao Zhang
# Date:   2021-02-05

from collections import defaultdict
class RandomizedSet:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.dic = defaultdict(int)
        self.vals = []

    def insert(self, val: int) -> bool:
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        """
        if val in self.dic:
            return False
        self.dic[val] = len(self.vals)
        self.vals.append(val)
        return True

    def remove(self, val: int) -> bool:
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        """
        if val not in self.dic:
            return False
        idx = self.dic[val]
        self.dic[self.vals[-1]] = idx
        self.dic.pop(val)
        self.vals[idx], self.vals[-1] = self.vals[-1], self.vals[idx]
        self.vals.pop()
        return True

    def getRandom(self) -> int:
        """
        Get a random element from the set.
        """
        idx = random.randint(0, len(self.vals) - 1)
        return self.vals[idx]


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
